src/math/lp/lar_solver.cpp

```diff
@@ -1057,6 +1034,110 @@ namespace lp {
         }
     }
 
+    u_dependency* lar_solver::get_dependencies_of_maximum(const lar_term& term) {        
+        const auto& s = this->m_mpq_lar_core_solver.m_r_solver;
+        // The sum of m_d[j]*x[j] = term.
+        // Every j with positive m_d[j] is at its upper bound,
+        // and every j with negative m_d[j] is at its lower bound: so the sum cannot be increased.
+        // All variables j in the sum are non-basic.
+        u_dependency* dep = nullptr;
+        for (unsigned j = 0; j < this->m_mpq_lar_core_solver.m_n(); j++) {
+            if (s.m_basis_heading[j] >= 0) {
+                SASSERT(s.m_d[j].is_zero());
+                continue;
+            }
+            const mpq& d_j = s.m_d[j];
+            if (d_j.is_zero()) continue;
+            TRACE("lar_solver", tout << "d[" << j << "] = " << d_j << "\n";);
+            TRACE("lar_solver", s.print_column_info(j, tout););
+            const ul_pair& ul = m_columns_to_ul_pairs[j];
+            svector<constraint_index> deps;    
+            u_dependency * bound_dep;
+            if (d_j.is_pos()) {
+                SASSERT(s.x_is_at_upper_bound(j));
+                bound_dep = ul.upper_bound_witness();
+              
+            } 
+            else {
+                SASSERT(s.x_is_at_lower_bound(j));
+                bound_dep = ul.lower_bound_witness();               
+            }
+            dep = m_dependencies.mk_join(dep, bound_dep);
+        }
+        return dep;
+    }
+
+    
+
+    void lar_solver::get_explanation_of_maximum(const lar_term& term, explanation& exp) {        
```

Comment:
seems you want the explain functionality to be there even in spite it is in a broken state with nested dependencies. 
There really is a dependency tree now. An explanation is a set of inequalities with coefficients. Each is justified by either a literal or, recursively, an explanation for the negation. 
When searching we just need the flattened set of literals, when justifying the inference it can be split into portions. You could also multiply the coefficients. For example, suppose the bound x <= k has coefficient c1, but is explained by the core x > k * c2 + bound2 * c3 + bound3 * c4.
you could include bound2 with coefficient c3*c1/c2, and bound3 with coefficient c4*c1/c2.

Reply:
I removed this method and the test method using explanation.